nifi-commons/nifi-security-utils/src/main/java/org/apache/nifi/security/pgp/StandardPGPOperator.java

```diff
@@ -0,0 +1,852 @@
+         if (hash != null && publicKeyCache.containsKey(hash)) {
+            return publicKeyCache.get(hash);
+        }
+
+        final PGPPublicKeys keys = new PGPPublicKeys();
+        JcaPGPPublicKeyRingCollection rings;
+
+        try {
+            InputStream decoderStream = PGPUtil.getDecoderStream(input);
+            rings = new JcaPGPPublicKeyRingCollection(decoderStream);
+            input.close();
+        } catch (final IOException | PGPException ignored) {
+            return null;
+        }
+
+        for (final PGPPublicKeyRing ring : rings) {
+            for (final PGPPublicKey key : ring) {
+                for (final Iterator<String> it = key.getUserIDs(); it.hasNext(); ) {
+                    keys.put(it.next(), key);
+                }
+            }
+        }
+
+        if (hash != null && keys.size() > 0) publicKeyCache.putMiss(hash, keys);
+        return keys;
+    }
+
+    /**
+     * Creates a list of public keys from the input stream.  When called with a lone {@link InputStream}, this method
+     * will not cache the resulting keys.
+     *
+     * @param input public key or key ring stream
+     * @return public keys
+     */
+    public PGPPublicKeys readPublicKeys(InputStream input) {
+        return readPublicKeys(input, null);
+    }
+
+    /**
+     * Creates a list of public keys from the given String.
+     *
+     * @param source public key text
+     * @return pubic keys
+     */
+    public PGPPublicKeys readPublicKeys(String source) {
+        return readPublicKeys(new ByteArrayInputStream(source.getBytes(StandardCharsets.UTF_8)), source.hashCode());
+    }
+
+    /**
+     * Creates a list of public keys from the given {@link File}.
+     *
+     * @param source public key File
+     * @return public keys
+     */
+    public PGPPublicKeys readPublicKeys(File source) throws FileNotFoundException {
+        return readPublicKeys(new FileInputStream(source), source.getAbsoluteFile().hashCode());
+    }
+
+    /**
+     * Creates a list of secret keys from the input stream.
+     *
+     * @param input secret key or key ring stream
+     * @return secret keys
+     */
+    PGPSecretKeys readSecretKeys(InputStream input, Integer hash) {
+        if (hash != null && secretKeyCache.containsKey(hash)) {
+            return secretKeyCache.get(hash);
+        }
+
+        final PGPSecretKeys keys = new PGPSecretKeys();
+        final KeyFingerPrintCalculator calc = new BcKeyFingerprintCalculator();
+        PGPSecretKeyRingCollection rings;
+
+        try {
+            rings = new PGPSecretKeyRingCollection(PGPUtil.getDecoderStream(input), calc);
+            input.close();
+        } catch (final IOException | PGPException ignored) {
+            return null;
+        }
+
+        for (final PGPSecretKeyRing ring : rings) {
+            for (final PGPSecretKey secretKey : ring) {
+                for (final Iterator<String> it = secretKey.getUserIDs(); it.hasNext(); ) {
```

Comment:
As with readPublicKeys this is not compatible with ECDSA/ECDH as the secret key for decryption is a sub key of the parent